aten/src/THC/THCTensorRandom.cuh

```diff
@@ -140,8 +140,16 @@ __device__ int binarySearchForMultinomial(T* dist,
 
```

Comment:
GitHub won't let me tag the right line, but could you also call this: https://github.com/colesbury/pytorch/blob/1a21d910e18a0104fc1f399add8f07b86d7cfdfe/aten/src/THC/THCTensorRandom.cu#L82-L88 on the result of `curand_uniform`?

Because the RNG returns (0, 1] I'm worried that we're skewing the probability distribution by just selecting the last non-zero element everytime a 1 is generated.

Reply:
Well, the code you point to selected 0 every time a 1 is generated. But actually, based on my limited understanding of the code, I think you do want to have (0,1] to be fed into the [binary search](https://github.com/pytorch/pytorch/blob/e40425fd9b7290cbd0f98667664f442fc032044b/aten/src/THC/THCTensorRandom.cuh#L130-L139) to yield correct results and would have to change the bisection details if 0 is to be fed into the search (which is actually something to look at for half).
I'm also not particularly concerned about the skew in sampling because you would likely need a **lot** of samples before you have a statistically significant deviation.

Best regards

Thomas

Reply:
True. I tried sampling a lot and a got an average of 3 ones per 1 million samples.

Reply:
@zou3519 Btw, the code you linked is also skewing the distribution somewhat. The result isn't exactly uniform.